pkg/webhook/elasticquota/quota_topology.go

```diff
@@ -0,0 +1,475 @@
+func (qt *quotaTopology) fillQuotaDefaultInformation(quota *v1alpha1.ElasticQuota) error {
+	if quota.Labels == nil {
+		quota.Labels = make(map[string]string)
+	}
+	if quota.Annotations == nil {
+		quota.Annotations = make(map[string]string)
+	}
+
+	if parentName, exist := quota.Labels[extension.LabelQuotaParent]; !exist || len(parentName) == 0 {
+		quota.Labels[extension.LabelQuotaParent] = extension.RootQuotaName
+		klog.V(5).Infof("fill quota %v parent as root", quota.Name)
+	}
+	maxQuota, err := json.Marshal(&quota.Spec.Max)
+	if err != nil {
+		return fmt.Errorf("fillDefaultQuotaInfo marshal quota max failed:%v", err)
+	}
+	if sharedWeight, exist := quota.Annotations[extension.AnnotationSharedWeight]; !exist || len(sharedWeight) == 0 {
+		quota.Annotations[extension.AnnotationSharedWeight] = string(maxQuota)
+		klog.V(5).Infof("fill quota %v sharedWeight as max", quota.Name)
+	}
+	return nil
+}
+
+// validateQuotaTopology checks the quotaInfo's topology with its parent and its children (when update calls the function).
+func (qt *quotaTopology) validateQuotaTopology(quotaInfo *core.QuotaInfo) error {
+	// if the quotaInfo's parent is root and its IsParent is false, the following checks will be true, just return nil.
+	if quotaInfo.ParentName == extension.RootQuotaName && !quotaInfo.IsParent {
+		return nil
+	}
+
+	if err := qt.checkParentQuotaInfoExist(quotaInfo.Name, quotaInfo.ParentName); err != nil {
+		return err
+	}
+
+	if err := qt.checkSubAndParentGroupMaxQuotaKeySame(quotaInfo); err != nil {
+		return err
+	}
+
+	if err := qt.checkMaxQuota(quotaInfo); err != nil {
+		return err
+	}
+
+	if err := qt.checkMinQuotaSum(quotaInfo); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (qt *quotaTopology) checkParentQuotaInfoExist(quotaName, parentName string) error {
+	if parentName != extension.RootQuotaName {
+		parentInfo, find := qt.quotaInfoMap[parentName]
+		if !find {
+			return fmt.Errorf("%v has parentName %v but not find parentInfo in quotaInfoMap", quotaName, parentName)
+		}
+		if _, exist := qt.quotaHierarchyInfo[parentName]; !exist {
+			return fmt.Errorf("%v has parentName %v but not find parentInfo in quotaHierarchyInfo", quotaName, parentName)
+		}
+		if !parentInfo.IsParent {
+			return fmt.Errorf("%v has parentName %v but the parentQuotaInfo's IsParent is false", quotaName, parentName)
+		}
+	}
+	return nil
+}
+
+func (qt *quotaTopology) checkSubAndParentGroupMaxQuotaKeySame(quotaInfo *core.QuotaInfo) error {
+	if quotaInfo.ParentName != extension.RootQuotaName {
+		parentInfo := qt.quotaInfoMap[quotaInfo.ParentName]
+		if !checkQuotaKeySame(parentInfo.CalculateInfo.Max, quotaInfo.CalculateInfo.Max) {
+			return fmt.Errorf("checkSubAndParentGroupMaxQuotaKeySame failed: %v's key is not the same with %v",
+				quotaInfo.ParentName, quotaInfo.Name)
+		}
+	}
+
+	children, find := qt.quotaHierarchyInfo[quotaInfo.Name]
+	if !find || len(children) == 0 {
```

Comment:
!find not equal to len(children) == 0, should return err; and quotaHierarchyInfo is your inner structure, I don't think is need to check it, also in "checkParentQuotaInfoExist"

Reply:
fixed

Reply:
after discuss, keep the logic